packages/security/src/types.ts

```diff
@@ -0,0 +1,183 @@
+// Node module: @loopback/security
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+/**
+ * A symbol for stringified id of security related objects
+ */
+export const securityId = Symbol('securityId');
+
+/**
+ * Represent a user, an application, or a device
+ */
+export interface Principal {
+  /**
+   * Name/id
+   */
+  [securityId]: string;
+
+  // tslint:disable-next-line:no-any
+  [attribute: string]: any;
+}
+
+export class TypedPrincipal implements Principal {
+  constructor(
+    public readonly principal: Principal,
+    public readonly type: string,
+  ) {}
+  get [securityId]() {
+    return `${this.type}:${this.principal[securityId]}`;
+  }
+}
+
+export interface UserProfile extends Principal {
+  // basic user information
+}
+
+export interface Organization extends Principal {}
+
+export interface Team extends Principal {}
+
+export interface ClientApplication extends Principal {}
+
+export interface Role extends Principal {}
+
+/**
+ * Security attributes used to authenticate the subject. Such credentials
+ * include passwords, Kerberos tickets, and public key certificates.
+ */
+export interface Credential {}
+
+/**
+ * `Permission` defines an action/access against a protected resource. It's
+ * the `what` for security.
+ *
+ * There are three levels of permissions
+ *
+ * - Resource level (Order, User)
+ * - Instance level (Order-0001, User-1001)
+ * - Property level (User-0001.email)
+ *
+ * @example
+ * - create a user (action: create, resource type: user)
+ * - read email of a user (action: read, resource property: user.email)
+ * - change email of a user (action: update, resource property: user.email)
+ * - cancel an order (action: delete, resource type: order)
+ */
+export class Permission {
+  /**
+   * Action or access of a protected resources, such as `read`, `create`,
+   * `update`, or `delete`
+   */
+  action: string;
+
+  /**
+   * Type of protected resource, such as `Order` or `Customer`
+   */
+  resourceType: string;
+  /**
+   * Property of a protected resource type/instance, such as `email`
+   */
+  resourceProperty?: string;
+  /**
+   * Identity of a protected resource instance, such as `order-0001` or
+   * `customer-101`
+   */
+  resourceId?: string;
+
+  get [securityId]() {
+    const resIds: string[] = [];
+    resIds.push(this.resourceType);
+    if (this.resourceProperty) {
+      resIds.push(this.resourceProperty);
+    }
+    const inst = this.resourceId ? `:${this.resourceId}` : '';
+
+    return `${resIds.join('.')}:${this.action}${inst}`;
+  }
+}
+
+/**
+ * oAuth 2.0 scope
+ */
+export interface Scope extends Permission {
+  name: string;
+}
+
+/**
+ * `Subject` represents both security state and operations for a single
+ * request. It's the `who` for security.
+ *
+ * Such operations include:
+ * - authentication (login)
+ * - authorization (access control)
+ * - session access
+ * - logout
+ *
+ */
+export interface Subject {
```

Comment:
@raymondfeng should we make the `principals` more strict? Like it can only contain one item with `{type: 'USER'}`(same for other types), since theoretically one request is only from one user/app/device

Reply:
Like an object as follows?
```ts
{
  user?: Principal,
  application?: Principal,
  device?: Principal, 
  [other:string]: Principal,
}
```